### DISCUSSION TITLE
Question about error handling with custom schemas

### DISCUSSION THREAD

**Discussion by hbgl** [colinhacks/zod#1943] [Q&A] [answered]
Zod provides a method to define custom schemas via `z.custom()` but I have not found a straightforward way to report any errors / issues. Here's an example:

```ts
const f = z.custom<string[]>(arg => {
    if (typeof arg !== 'string') {
        // How to report errors / issues?
        return z.NEVER;
    }
    return arg.split(',');
});
```

From the [example](https://zod.dev/?id=custom-schemas) in the docs it seems that `z.custom()` can only validate but not parse. I did find a workaround with using `transform` but it does not feel right.

```ts
const f = z.custom<string[]>().transform((arg: any, ctx) => {
    if (typeof arg !== 'string') {
        ctx.addIssue({ code: 'custom' });
        return z.NEVER;
    }
    return arg.split(',');
});
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const schema = z
    .string()
    .transform(value => value.split(','))
    .pipe(z.string().array())
type Data = z.infer<typeof schema>
// type Data = string[]

console.log(schema.safeParse('a,b,c').success) // true
console.log(schema.safeParse(['a', 'b', 'c']).success) // false
```

**Comment by hbgl** [comment]
That'll work for this simple example. When more involved parsing is needed, then using `transform` would probably be appropriate. I initially thought `z.custom<T>` was a general purpose way to write custom parsers but it is actually just meant for edge cases like template strings.
